meetings/working-groups/params-improvements/PI-2022-11-03.md

```diff
@@ -0,0 +1,126 @@
+# Params Improvements Working Group Meeting for November 3, 2022
+The second meeting of the _`params` improvements_ working group was focused on the code generated for arguments to `params ReadOnlySpan<T>`. 
+
+## Span allocation
+The compiler _may allocate a span on the stack_ when:
+- the span is implicitly allocated,
+- the span length is known at compile time, and
+- the compiler can determine _through escape analysis_ that no references to the span escape the containing method.
+
+For arguments to a `params ReadOnlySpan<T>` parameter (which is implicitly `scoped`), the conditions above are satisfied.
+
+We have a request with the runtime team to support _fixed size arrays_ of managed types, at least for fields.
+If we have _fixed size array_ fields, we can define `struct` types with inline arrays and use locals for stack allocated arrays.
+
+For example, consider a `FixedSizeArray2` type defined below which includes an inline 2 element array:
+```csharp
+struct FixedSizeArray2<T>
+{
+    public T[2] Array; // pseudo-code for inline array
+}
+```
+
+With that type, a call to `WriteLine("{0}, {1}", x, y)` could be emitted as:
+```csharp
+var _tmp1 = new FixedSizeArray2<object>();
+_tmp1.Array[0] = x;
+_tmp1.Array[1] = y;
+var _tmp2 = new ReadOnlySpan<object>(_tmp.Array);
+
+// WriteLine(string format, params ReadOnlySpan<object> args);
+WriteLine("{0}, {1}", _tmp2);
+```
+
+Ideally the BCL will provide types such as `FixedSizeArray1`, `FixedSizeArray2`, etc. for a limited number of array lengths.
+And if the compilation requires spans for other argument lengths, the compiler would generate and emit the additional types.
+
+The compiler will implicitly allocate `params` argument spans on the stack regardless of the number of arguments.
+To avoid implicit stack allocation at a call site, the calling code should allocate the span explicitly.
```

Comment:
oh, i did'nt realize we made that determination.  this seems risky, if we are going to just add all these new apis.  someone could upgrade, recompile, and potentially blow stack.  e.g. a call to some string.format that takes 1000 args, right?

Reply:
I don't think this is the full picture. I expected that we would also request a JIT feature where large fixed-size arrays can be heap-allocated instead if heuristics owned by the runtime determine that's the right thing to do.

Reply:
I think we need justification for worrying about that scenario. If we really want large fixed size locals to be heap allocated we need concrete evidence this is a real problem. Not speculation. I know I said we *could* do this. However, we won't build such a heuristic without evidence, as we won't be able to build a good heuristic without customer experience, as this sort of magic behavior is likely have negative effects as well. In applications that I've seen, deep stacks are very much in the minority, and applications could easily tolerate a 1000 arg stack allocation. (Without strong evidence, I believe we would declare the heuristic to be .... we always allocate on the stack.)

Reply:
@davidwrighton i'm ok with that approach.  But i do think we need to be careful here as we're talking about a silent source-change for existing apps.

Reply:
It could be that the cases where people risk using up the stack are usually advanced scenarios already. Given that, it wouldn't necessarily be that difficult for people to identify the particularly large stack frames and replace `M(lots, of, args)` with `M(new[] { lots, of, args })` where needed.